### PR TITLE
apiserver/filters test: fix data race and do not leak goroutines

### PR DESCRIPTION
**Pull request by adtac**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**: The test contains a data race that was triggered due to flakiness elsewhere ([example](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/96124/pull-kubernetes-bazel-test/1324042663611076608)). The test also leaked goroutines and did not wait for all goroutines to finish. This PR fixes both issues.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/sig api-machinery
/cc @MikeSpreitzer @bjrara @ahg-g

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go [hunk 1]
```diff
@@ -294,14 +294,24 @@ func TestApfExecuteMultipleRequests(t *testing.T) {
...
+	var err error
 	for i := 0; i < concurrentRequests; i++ {
-		var err error
+		wg.Add(1)
 		go func() {
-			err = expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK)
+			errLocal := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK)
+			if errLocal != nil {
+				mu.Lock()
+				err = errLocal
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:306]: if multiple fail, you'll only see one. Do you want to least `t.Log` all of them?

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go [hunk 2]
```diff
@@ -294,14 +294,24 @@ func TestApfExecuteMultipleRequests(t *testing.T) {
...
+	var wg sync.WaitGroup
+	var mu sync.Mutex
+	var err error
 	for i := 0; i < concurrentRequests; i++ {
-		var err error
+		wg.Add(1)
 		go func() {
-			err = expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK)
+			errLocal := expectHTTPGet(fmt.Sprintf("%s/api/v1/namespaces/default", server.URL), http.StatusOK)
+			if errLocal != nil {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:304]: Is `t` not threadsafe?

```go
if err != nil{
  t.Error(err)
}
```


Seems much simpler if it is.

Comment (adtac) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go]: Indeed, much simpler! I can't believe I missed that :D I confused Error (which is concurrency safe) with Fatal, etc. (which isn't concurrency-safe):

> A test ends when its Test function returns or calls any of the methods FailNow, Fatal, Fatalf, SkipNow, Skip, or Skipf. Those methods, as well as the Parallel method, must be called only from the goroutine running the Test function.

>The other reporting methods, such as the variations of Log and Error, may be called simultaneously from multiple goroutines.

https://golang.org/pkg/testing/#T

Done.

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go (hunk 1 above)
Comment (adtac) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go]: Done.

**Review by MikeSpreitzer — APPROVED** [review]
/LGTM
Sorry, I should have caught these.

**Comment by yue9944882** [comment]
dang! sry i missed it

**Comment by MikeSpreitzer** [comment]
Let's merge #96244 instead of this one.

**Comment by deads2k** [comment]
> Let's merge #96244 instead of this one.

I think that PR can separately stand on its own merit, but I'm not ready to eliminate retries b/c I don't yet know what they are there for.

/approve

**Comment by MikeSpreitzer** [comment]
@deads2k : note that we are talking about retries in a test.  A test that passes without the retries.

**Comment by bjrara** [comment]
@adtac Thanks for helping fix the issue.

**Pull request merged**